### PR TITLE
Enable gradient styles on the legend items

### LINKED ISSUE #28: Legend boxes don't reflect the gradient
### Feature Proposal

As I have seen in my lib, when you use  gradients, the gradients are not reflected in the legend.

To be precise, the gradient is applied to the legend but due to the gradient size is limited to the scales, the legend are out side of the gradient and then the canvas use the color related to the position.

See example in codepen: https://codepen.io/stockinail/pen/xxpENVM

As you can see the legend is using the `red` color because is the color from the gradient for legend position.

The same is for polarArea (using the radial gradient) where all legend items are colored with the color of the gradient.

polarGradient

The proposal is to reflect on the legend the gradient.

### Possible Implementation 

I see only 1 option, creating a `legendLabels` callback in the plugin to set to the chart in order to create a specific gradient (with dimension of the legend box). 

For polarArea, the `legendLabels` callback could return the color at the data index, calculating the last color for the data index. For this, I have got already a code (even if is written in [Java](https://github.com/pepstock-org/Charba/blob/master/src/org/pepstock/charba/client/colors/Gradient.java#L223-L280)) that it can be translated and used.

### PR DESCRIPTION
**Pull request by stockiNail**
Fix #28 

This PR is enabling to set `fillStyle` and `strokeStyle` on the legend items related to datasets where the gradient has been set.

It is applying only `backgroundColor` and `borderColor`, if set.

<table>
  <tr>
    <td>
      
    </td>
  </tr>
</table> 
It doesn't go in exception on CHART.JS version 2.x but it can not enable the styles on legend items because the legend size calculation is done in a different time of chart lifecycle.

### TODO

- [x] enable at dataset level
- [x] enable at data level (for polarArea)
- [x] test cases scale 0-positive (covered)
- [x] test cases scale negative-0
- [x] test cases scale positive-positive
- [x] test cases scale negative-positive
- [x] test cases scale negative-negative

### CODE REVIEW

File: src/colors.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import {color} from 'chart.js/helpers';
+import {getGradientData} from './helpers';
+
+const rgbs = (c) => Math.round(c._rgb.a * 255) << 24 | c._rgb.r << 16 | c._rgb.g << 8 | c._rgb.b;
+// IEC 61966-2-1:1999
+const toRGBs = (l) => l <= 0.0031308 ? l * 12.92 : Math.pow(l, 1.0 / 2.4) * 1.055 - 0.055;
+// IEC 61966-2-1:1999
+const fromRGBs = (srgb) => srgb <= 0.04045 ? srgb / 12.92 : Math.pow((srgb + 0.055) / 1.055, 2.4);
+
+function interpolate(percent, startColor, endColor) {
```
Comment (kurkle) [inline src/colors.js:10]: Is this equivalent to `color(startColor).mix(endColor, percent)`?

Comment (stockiNail) [reply src/colors.js:10]: > Is this equivalent to `color(startColor).mix(endColor, percent)`?

I didn't check that color function... let me try! it will simplify a lot!

Comment (stockiNail) [reply src/colors.js:10]: I doesn't seem it's working as I'm expecting:

color(startColor).mix(endColor, percent)

colorMix

function interpolate(percent, startColor, endColor) {

colorNoMix

As you can see, it seems mix is not precise enough.

Comment (kurkle) [reply src/colors.js:10]: Oh right, its not the same.

There are few things in the color lib that you can use though. like set/get the rgb property of a color.
I think I'll add that `interpolate` to it :)

Comment (stockiNail) [reply src/colors.js:10]: FINE! I have got that code in my lib as well and is based to IEC 61966-2-1:1999 ;)

File: src/colors.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  const startRGBs = rgbs(startColor.color);
+  const startR = fromRGBs(((startRGBs >> 16) & 0xff) / 255);
+  const startG = fromRGBs(((startRGBs >> 8) & 0xff) / 255);
+  const startB = fromRGBs((startRGBs & 0xff) / 255);
```
Comment (kurkle) [inline src/colors.js:14]: No need to convert to int and back?

```suggestion
  const start = startColor.color.rgb;
  const startR = fromRGBs(start.r / 255);
  const startG = fromRGBs(start.g / 255);
  const startB = fromRGBs(start.b / 255);
```

File: src/colors.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  const endRGBs = rgbs(endColor.color);
+  const endR = fromRGBs(((endRGBs >> 16) & 0xff) / 255);
+  const endG = fromRGBs(((endRGBs >> 8) & 0xff) / 255);
+  const endB = fromRGBs((endRGBs & 0xff) / 255);
+
+  const r = Math.round(toRGBs(startR + percent * (endR - startR)) * 255);
+  const g = Math.round(toRGBs(startG + percent * (endG - startG)) * 255);
+  const b = Math.round(toRGBs(startB + percent * (endB - startB)) * 255);
+
+  return color({r, g, b, a: 1});
```
Comment (kurkle) [inline src/colors.js:25]: Not sure if alpha channel can be used with gradients, but if it can, it should be interpolated too.

Something like 
```js
const a = startB + percent * (endB - startB)
```

Another note, can do the calculation while constructing the result object (to avoid creating intermediate variables):
```js
return color({
  r: Math.round(toRGBs(startR + percent * (endR - startR)) * 255),
  ...
});
```

File: src/colors.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * @param {Object} state - state of the plugin
+ * @param {{key: string, legendItemKey: string}} keyOption - option of the dataset where the gradient is applied
+ * @param {number} datasetIndex - dataset index
+ * @param {number} value - value used for searching the color
+ * @returns {Object} calculated color
+ */
+export function getInterpolatedColorByValue(state, keyOption, datasetIndex, value) {
+  const data = getGradientData(state, keyOption, datasetIndex);
+  if (!data) {
+    return;
```
Comment (kurkle) [inline src/colors.js:39]: can this happen?

File: src/colors.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  }
+  const percent = data.scale.options.reverse ? 1 - value / data.scale.max : value / data.scale.max;
```
Comment (kurkle) [inline src/colors.js:41]: this assumes the scale starts from zero?

File: src/index.js [hunk 6]
```diff
@@ -58,19 +58,33 @@ function setValue(meta, dataset, key, value) {
...
+    state.options.set(key, stateOptions);
+  }
+  return stateOptions;
+}
 
 export default {
   id: 'gradient',
+
+  beforeInit(chart) {
+    const state = new Object();
```
Comment (kurkle) [inline src/index.js:77]: So JAVA :)

`new Object()` is equivalent to `{}`, except its probably slower.

File: src/index.js [hunk 7]
```diff
@@ -88,12 +102,29 @@ export default {
...
+          applyColors(value, option.stopColors);
           setValue(meta, dataset, key, value);
         }
       }
     }
+  },
+
+  afterUpdate(chart) {
+    const state = chartStates.get(chart);
+    if (chart.legend && Chart.version) {
```
Comment (kurkle) [inline src/index.js:125]: Because the Chart.version is starting to appear more frequently, should add a constant about it at the top:

```js
const isChartV3 = Chart.version;
```

**Review by kurkle — COMMENTED** [review]
Impressive!

I had some comments and I think some more tests with scales not starting from zero.
I think we need to cover all these scale cases:
0-positive (covered)
negative-0
positive-positive
negative-positive
negative-negative

**Comment by stockiNail** [comment]
> Impressive!
> 
> I had some comments and I think some more tests with scales not starting from zero. I think we need to cover all these scale cases: 0-positive (covered) negative-0 positive-positive negative-positive negative-negative

Yes, I put it in Draft because there a couple of things where I have to pay attention:

1. change the calculation of percentage using `getPixelStop` (unique way everywhere)
2. understand if the `afterDraw` hook is the right one to change the legend. I think it could be a problem if the animation is disabled
3. destroy the state when the chart is destroying

This afternoon I didn't have time to go on. I'll do asap.

File: src/colors.js (hunk 3 above)
Comment (stockiNail) [reply src/colors.js:25]: > Not sure if alpha channel can be used with gradients, but if it can, it should be interpolated too.

If you see some previous commits, I was calculating but if you apply the alpha, it seems the legend is not showing it. I'll do some additional tests.


> Another note, can do the calculation while constructing the result object (to avoid creating intermediate variables):

OKI!!!

File: src/colors.js (hunk 4 above)
Comment (stockiNail) [reply src/colors.js:39]: I have to check but if you configure the gradient for dataset options which are not part of the legend rendering (i.e. hoverBackgroundColor), the result could be inconsistent.

File: src/index.js (hunk 6 above)
Comment (stockiNail) [reply src/index.js:77]: > So JAVA :)

Yes... be patient but sometimes I forgot that I'm developing in JS ;)

File: src/index.js (hunk 7 above)
Comment (stockiNail) [reply src/index.js:125]: OK!

File: src/index.js (hunk 6 above)
Comment (stockiNail) [reply src/index.js]: done

File: src/colors.js (hunk 5 above)
Comment (stockiNail) [reply src/colors.js]: move getPixelStop  and tests with all scale values

File: src/colors.js (hunk 3 above)
Comment (stockiNail) [reply src/colors.js]: I have done some tests but I think it's not correct to calculate the interpolation on the alpha, based on percentage of the value, because when the value is small, the alpha is close to zero and the color is not showed. Maybe the max of alphas could be calculated. If you agree, I'll remain at `a: 1`

Comment (stockiNail) [reply src/colors.js]: I was wrong... added interpolation on alpha as well.

Comment (stockiNail) [reply src/colors.js]: Thanks @kurkle because you gave me the input to change my lib as well to be more performed!!

File: src/colors.js (hunk 4 above)
Comment (stockiNail) [reply src/colors.js:39]: I have added test cases where only `hoverBorderColor` has been set and the data is undefined

**Comment by stockiNail** [comment]
@kurkle 41 files.... I know you don't like big PR. Sorry for that.

File: src/index.js [hunk 8]
```diff
@@ -58,42 +37,80 @@ function setValue(meta, dataset, key, value) {
...
+    stateOptions = stateOptions.filter((el) => el.datasetIndex !== datasetIndex);
+    state.options.set(key, stateOptions);
+  }
+  return stateOptions;
+}
 
-const areaIsValid = (area) => area && area.right > area.left && area.bottom > area.top;
+function updateDataset(chart, state, gradient, dataset, datasetIndex) {
+  const ctx = chart.ctx;
+  const meta = chart.getDatasetMeta(datasetIndex);
```
Comment (kurkle) [inline src/index.js:54]:
```suggestion
  const meta = chart.getDatasetMeta(datasetIndex);
  if (meta.hidden) {
    return;
  }
```

File: src/index.js [hunk 9]
```diff
@@ -58,42 +37,80 @@ function setValue(meta, dataset, key, value) {
...
+  for (const [key, options] of Object.entries(gradient)) {
+    const {axis, colors} = options;
+    const scale = getScale(meta, axis);
```
Comment (kurkle) [inline src/index.js:57]:
```suggestion
    if (!colors) {
      continue;
    }
    const scale = getScale(meta, axis);
```

File: src/index.js [hunk 10]
```diff
@@ -58,42 +37,80 @@ function setValue(meta, dataset, key, value) {
...
+        stopColors: []
+      };
+      stateOptions.push(option);
+      const value = createGradient(ctx, axis, scale);
+      addColors(scale, colors, option.stopColors);
+      if (option.stopColors.length) {
+        applyColors(value, option.stopColors);
+        setValue(meta, dataset, key, value);
+      }
+    }
```
Comment (kurkle) [inline src/index.js:77]:
```suggestion
    const option = {
      datasetIndex,
      axis,
      scale,
      stopColors: []
    };
    stateOptions.push(option);
    const value = createGradient(ctx, axis, scale);
    addColors(scale, colors, option.stopColors);
    if (option.stopColors.length) {
      applyColors(value, option.stopColors);
      setValue(meta, dataset, key, value);
    }
```

**Review by kurkle — APPROVED** [review]
Some suggestions, that don't really make any difference, but could save a couple of cpu cycles. :)

I'm not sure about the state handling, but that is internals and can be changed at any time. So its ok.

**Comment by stockiNail** [comment]
> I'm not sure about the state handling, but that is internals and can be changed at any time.

OK. I can review later. I tought about the structure, using map. And also an another issue was that even if the dataset is hidden, the state should be present for the legend.
